File: backend/src/sources/probuilds.js
import { default_schema } from '../shared/index.js';
import { cl, trinksCon } from '../helpers/index.js';
import ChampionifyErrors from '../errors.js';
import T from '../translate.js';
import progressbar from '../progressbar.js';
import store from '../store.js';

class WebLogger {
  error(...a) { console.error('[Championify]', ...a); }
  info(...a) { console.info('[Championify]', ...a); }
  warn(...a) { console.warn('[Championify]', ...a); }
}
const logger = new WebLogger();

export const source_info = {
  name: 'ProBuilds',
  id: 'probuilds',
};

async function getChamps() {
  const res = await fetch(
    `https://blitz-cdn-plain.blitz.gg/blitz/ddragon/${await getVersion()}/data/en_US/champions.json`
  );
  const data = await res.json();
  const champions = data?.champions || [];
  return champions.map(champ => {
    const id = champ.key;
    const name = champ.name;
    store.set(`champ_ids.${id}`, id);
    return { id, name };
  });
}

async function getItems(champion, position) {
  try {
    cl(`${T.t('processing')} ProBuilds: ${T.t(champion.name)} - ${T.t(position)}`);
  } catch (_error) {
    store.push('undefined_builds', {
      source: source_info.name,
      champ: champion,
      position,
    });
    throw new ChampionifyErrors.Error('ProBuilds : ' + _error);
  }

  const variables = {
    championId: champion.id,
    queue: 'RANKED_SOLO_5X5',
    role: position,
  };

  try {
    const response = await fetch('https://datalake.v2.iesdev.com/graphql', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        query: `
          query ChampionBuildsTagged($championId:String! $queue:String! $role:String) {
            executeDatabricksQuery(
              game: LEAGUE
              queryName: "prod_champion_builds_tags"
              params: [
                {name: "individual_position", value: $role}
                {name: "queue_id", value: $queue}
                {name: "champion_id", value: $championId}
              ]
            ) {
              payload
            }
          }
        `,
        variables,
      }),
    });

    const json = await response.json();
    const dataArray = json?.data?.executeDatabricksQuery?.payload?.result?.dataArray;
    if (!dataArray) throw new Error('No builds data found');

    const itemBuildsJson = dataArray?.[0]?.[8];
    let itemBuilds = [];
    if (itemBuildsJson) {
      try {
        itemBuilds = JSON.parse(itemBuildsJson);
      } catch (e) {
        logger.warn(`Failed to parse item builds for ${champion.id}: ${e}`);
      }
    }

    const blocks = itemBuilds.map(build => ({
      items: build.itemIds.split(',').map(id => ({ id, count: 1 })),
      type: 'Most Popular',
    }));

    const riot_json = {
      ...default_schema,
      champion: champion.id,
      title: `ProBuilds ${dataArray?.[0]?.[0] ?? ''}`,
      blocks: trinksCon(blocks),
    };

    progressbar.incrChamp();
    return {
      champ: champion,
      file_prefix: 'all',
      riot_json,
      source: source_info.id,
    };
  } catch (err) {
    logger.error(err);
    store.push('undefined_builds', {
      source: source_info.name,
      champ: champion,
      position: 'All',
    });
  }
}

export async function getSr() {
  const positions = ['TOP', 'JUNGLE', 'MID', 'BOT', 'SUPPORT'];
  const champions = await getChamps();
  const promises = [];
  for (const champion of champions) {
    for (const position of positions) {
      promises.push(getItems(champion, position));
    }
  }
  const results = await Promise.all(promises);
  const data = results.filter(v => v != null);
  store.push('sr_itemsets', data);
}

export async function getVersion() {
  try {
    const r = await fetch('https://utils.iesdev.com/static/json/lol/riot/versions', {
      headers: { 'Content-Type': 'application/json' },
    }).then(res => res.json());
    return r?.[0] || '';
  } catch (err) {
    logger.error(err);
    return '';
  }
}

File: src-tauri/src/scrapers/probuilds.rs
use std::time::Duration;

use async_trait::async_trait;
use reqwest::Client;
use serde_json::{json, Value};
use anyhow::{Result, Context};
use crate::{error::AppError, riot_cdn::get_riot_champions, scrapers::{BuildScraper, ScraperContext, registry::ScraperHealth}};

use super::{SourceInfo, ItemBlock, Item, RiotJsonSR, BuildResult, Champion};

pub struct ProBuildsScraper;

#[async_trait]
impl BuildScraper for ProBuildsScraper {

    fn source_info(&self) -> SourceInfo {
        SourceInfo {
            name: "ProBuilds".to_string(),
            id: "probuilds".to_string(),
        }
    }

    async fn get_version(&self, client: &Client) -> Result<String, AppError> {
        let response = client
            .get("https://utils.iesdev.com/static/json/lol/riot/versions")
            .header("Content-Type", "application/json")
            .timeout(Duration::from_secs(10))
            .send()
            .await?;
        
        let versions: Vec<String> = response.json().await?;
        Ok(versions.first().cloned().unwrap_or_default())
    }

    async fn get_sr(&self, client: &Client, context: &ScraperContext, _role: Option<&str>) -> Result<Vec<BuildResult>, AppError> {
        let positions = ["TOP", "JUNGLE", "MID", "BOT", "SUPPORT"];
        let version = context.riot_version.clone();
        let champions = get_riot_champions(client, &version).await?;
        
        let mut results = Vec::new();
        
        for champion_value in champions.as_array().context("Champions is not an array")? {
            let champion: Champion = match serde_json::from_value(champion_value.clone()) {
                Ok(champ) => champ,
                Err(e) => {
                    eprintln!("Failed to parse champion: {}", e);
                    continue;
                }
            };

            for position in &positions {
                match get_items(client, &champion, position).await {
                    Ok(build) => results.push(build),
                    Err(e) => eprintln!("Error for {} {}: {}", champion.name, position, e),
                }
            }
        }
        
        Ok(results)
    }
    async fn check_health(&self, client: &Client) -> ScraperHealth {
        match self.get_version(client).await {
            Ok(_) => ScraperHealth::Available,
            Err(_) => ScraperHealth::Down,
        }
    }
    async fn needs_riot_version(&self) -> bool {
        false
    }
}

async fn get_items(client: &Client, champion: &Champion, position: &str) -> Result<BuildResult, AppError> {
    println!("Processing ProBuilds: {} - {}", champion.name, position);
    
    let query = r#"
        query ChampionBuildsTagged($championId:String! $queue:String! $role:String) {
            executeDatabricksQuery(
                game: LEAGUE
                queryName: "prod_champion_builds_tags"
                params: [
                    {name: "individual_position", value: $role}
                    {name: "queue_id", value: $queue}
                    {name: "champion_id", value: $championId}
                ]
            ) {
                payload
            }
        }
    "#;
    
    let variables = json!({
        "championId": champion.id,
        "queue": "RANKED_SOLO_5X5",
        "role": position,
    });
    
    let response = client
        .post("https://datalake.v2.iesdev.com/graphql")
        .header("Content-Type", "application/json")
        .json(&json!({
            "query": query,
            "variables": variables,
        }))
        .send()
        .await?;
    
    let json: Value = response.json().await?;
    if let Some(errors) = json.get("errors") {
        return Err(AppError::Custom(format!("GraphQL errors: {}", errors)));
    }
    let data_array = json
        .pointer("/data/executeDatabricksQuery/payload/result/dataArray")
        .and_then(|v| v.as_array())
        .context("No builds data found")?;
    
    let item_builds_json = data_array
        .get(0)
        .and_then(|arr| arr.get(8))
        .and_then(|v| v.as_str())
        .unwrap_or("[]");
    
    let item_builds: Vec<Value> = serde_json::from_str(item_builds_json).unwrap_or_default();
    
    let blocks: Vec<ItemBlock> = item_builds
        .iter()
        .filter_map(|build| {
            let item_ids = build.get("itemIds")?.as_str()?;
            let items: Vec<Item> = item_ids
                .split(',')
                .map(|id| Item {
                    id: id.to_string(),
                    count: 1,
                })
                .collect();
            
            Some(ItemBlock {
                block_type: "Most Popular".to_string(),
                items,
            })
        })
        .collect();
    
    let title = data_array
        .get(0)
        .and_then(|arr| arr.get(0))
        .and_then(|v| v.as_str())
        .unwrap_or("");
    
    Ok(BuildResult {
        champ: champion.name.clone(),
        file_prefix: "all".to_string(),
        riot_json: RiotJsonSR {
            champion: champion.id.clone(),
            title: format!("ProBuilds {}", title),
            blocks,
            map: None,
        },
        source: "probuilds".to_string(),
    })
}
